app/src/main/java/org/schabi/newpipe/streams/OggFromWebMWriter.java

```diff
@@ -0,0 +1,434 @@
+        }
+        if (!target.canWrite() || !target.canRewind()) {
+            throw new IllegalArgumentException("output stream must be writable and allows seeking");
+        }
+
+        this.source = source;
+        this.output = target;
+
+        this.STREAM_ID = (int) System.currentTimeMillis();
+
+        populate_crc32_table();
+    }
+
+    public boolean isDone() {
+        return done;
+    }
+
+    public boolean isParsed() {
+        return parsed;
+    }
+
+    public WebMTrack[] getTracksFromSource() throws IllegalStateException {
+        if (!parsed) {
+            throw new IllegalStateException("source must be parsed first");
+        }
+
+        return webm.getAvailableTracks();
+    }
+
+    public void parseSource() throws IOException, IllegalStateException {
+        if (done) {
+            throw new IllegalStateException("already done");
+        }
+        if (parsed) {
+            throw new IllegalStateException("already parsed");
+        }
+
+        try {
+            webm = new WebMReader(source);
+            webm.parse();
+            webm_segment = webm.getNextSegment();
+        } finally {
+            parsed = true;
+        }
+    }
+
+    public void selectTrack(int trackIndex) throws IOException {
+        if (!parsed) {
+            throw new IllegalStateException("source must be parsed first");
+        }
+        if (done) {
+            throw new IOException("already done");
+        }
+        if (webm_track != null) {
+            throw new IOException("tracks already selected");
+        }
+
+        switch (webm.getAvailableTracks()[trackIndex].kind) {
+            case Audio:
+            case Video:
+                break;
+            default:
+                throw new UnsupportedOperationException("the track must an audio or video stream");
+        }
+
+        try {
+            webm_track = webm.selectTrack(trackIndex);
+        } finally {
+            parsed = true;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        done = true;
+        parsed = true;
+
+        webm_track = null;
+        webm = null;
+
+        if (!output.isClosed()) {
+            output.flush();
+        }
+
+        source.close();
+        output.close();
+    }
+
+    public void build() throws IOException {
+        float resolution;
+        SimpleBlock bloq;
+        ByteBuffer header = ByteBuffer.allocate(27 + (255 * 255));
+        ByteBuffer page = ByteBuffer.allocate(64 * 1024);
+
+        header.order(ByteOrder.LITTLE_ENDIAN);
+
+        /* step 1: get the amount of frames per seconds */
+        switch (webm_track.kind) {
+            case Audio:
+                resolution = getSampleFrequencyFromTrack(webm_track.bMetadata);
+                if (resolution == 0f) {
+                    throw new RuntimeException("cannot get the audio sample rate");
+                }
+                break;
+            case Video:
+                // WARNING: untested

```

Comment:
What does this mean?

Reply:
ogg can be used for muxing a video stream (normally THEORA), but i never tested. is here beacause only took me 3 lines of code :)